### ISSUE TITLE
Resource based policy for secrets manager secrets

### ISSUE THREAD

**Issue by hhurtado** [terraform-providers/terraform-provider-aws#5329]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
AWS Secrets Manager allows to attach, delete and retrieve a resource based policy which controls which principals have rights to operate on the secret.

It would be great to have the ability to create and retreive this policy in Terraform configurations.

### New or Affected Resource(s)

* aws_secretsmanager_secret_policy new resource
* aws_secretsmanager_secret datasource new policy attribute

### Potential Terraform Configuration

```hcl
resource "aws_secretsmanager_secret" "MySecret" {
  name = "database-credentials"
}

resource "aws_secretsmanager_secret_policy" "policy" {
  secret_id = "${aws_secretsmanager_secret.MySecret.id}"
  policy =<<POLICY
{
  "Version": "2012-10-17",
  "Id": "MYSECRETPOLICY",
  "Statement": [
    {
      "Sid": "RoleAllow",
      "Effect": "Allow",
      "Principal": {"AWS": ["arn:aws:iam::111122223333:role/myRole"]},
      "Action": "secretsmanager:GetSecretValue",
      "Resource": "*",
    } 
  ]
}
POLICY
}

data "aws_secretsmanager_secret" "byName" {
  name = "database-credentials"
}

output "secret_policy" {
  value = "${data.aws_secretsmanager_secret.byName.policy}"
}
```

### References
* https://docs.aws.amazon.com/cli/latest/reference/secretsmanager/put-resource-policy.html
* https://docs.aws.amazon.com/cli/latest/reference/secretsmanager/get-resource-policy.html

**Comment by bflad** [comment]
Hi @hhurtado 👋  Thanks for submitting this.

We actually have support for managing resource policies merged into the existing `aws_secretsmanager_secret` resource via #5290, which will release with version 1.29.0 of the AWS provider, likely later today. 👍 

As for the data source, that has not been implemented yet.

**Comment by hhurtado** [comment]
Great, Thanks a lot.

**Comment by tkbky** [comment]
Made a [PR](https://github.com/terraform-providers/terraform-provider-aws/pull/6091) for this.

**Comment by bflad** [comment]
Support for the `aws_secretsmanager_secret` data source `policy` attribute has been merged and will release with version 1.40.0 of the AWS provider, likely middle of this week. 👍

**Comment by bflad** [comment]
This has been released in [version 1.40.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1400-october-10-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
